[1단계 - 자동차 경주 구현] 차리(이찬주) 미션 제출합니다.

cjlee38 opened the pull request:
안녕하세요? 먼저 코드 리뷰에 대해서 감사의 말씀 드립니다. :)
프리코스때 한 번 진행한 적이 있었던 미션이긴 하지만,
페어와 다양한 논의를 하면서 진행하다보니,
이전 프리코스 때 고민했던 것들, 그리고 또 페어가 고민했던 것들이 여러가지 떠올랐는데요.
관련하여, 생각이 짧아 나름대로 내린 결론이 괜찮은지 아닌지에 대한 고민이 많아 질문 드립니다!

### 1. View와 Domain의 구분
기본적으로 View는 사용자의 UI를 담당하고, Domain은 실질적인 비즈니스 로직이 포함되어 있는 Scope 일 텐데요.
그러다보니 Input을 담당하는 View의 경우 input을 받아서 그대로(String) 돌려주기만 하고, controller는 이를 받아서
domain으로 전달하는 역할을 맡고, domain에서 실질적으로 객체를 생성하는 구조로 코드를 작성했습니다.

그런데 그러다보니, domain에서 검증로직도 포함되어야 하고, 
또 하위 객체를 실질적으로 생성하는 행위(e.g. 여러 Car를 관리하는 객체인 Cars의 경우,
 여러 Car를 만들기 위해 ,(쉼표)로 문자열을 구분하는 것) 도 있어야 하고,
비즈니스 로직을 적용하기도 하는(e.g. 자동차의 전진)데, 이러다 보니 **하나의 클래스가 너무 많은 일을
하고 있는것이 아닐까?** 하는 의문이 들었습니다.

이를 해결하려면 검증로직 혹은 생성로직을 별도의 클래스로 구분할 수도 있는데, 그렇게 될 경우
일급 컬렉션에서 이야기하는 **"상태와 행위를 한 곳에서 관리한다" 라는 측면에서 조금 어긋나는 것** 처럼 보였습니다.

결국 한 곳에 모으는 것으로 결정을 내리긴 하였으나... 이게 과연 괜찮은 것인지? 다른 방법이 있는데 제가 모르는 것인지... 가 궁금합니다.

### 2. 테스트 코드의 작성 범위.
이번 미션의 요구사항에는 모든 로직에 대하여 테스트 코드를 작성할 것이 요구되었는데요.
테스트 코드를 작성할 때 얼마나 자세히 테스트를 해야 하는지에 대한 고민입니다.

가령, 자동차 Car를 구성하는 이름의 Name, 위치의 Position이 있을 때, 
Name과 Position 내에서 올바른 입력이 주어졌는지 검증하는 로직이 있고, 이에 대한 테스트 코드도 작성했는데요.

그렇다면 Car의 생성에 대해서 검증하는 테스트가 과연 필요한지? 에 대한 의문이 들었습니다.
물론 "어떠한 작업 이후 Name과 Position을 생성한다" 라는 가정이라면 테스트가 필요하겠지만..
"곧바로 Name과 Position에게 위임해버리는 경우라면 하지 않아도 될까?" 하는 것이 의문입니다.

문제는 이러한 고민이 단순한 생성로직 뿐만 아니라 로직 메소드에도 적용된다는 점인데요.
자동차의 전진 전략을 담당하는 MovementStrategy 객체를, 자동차의 위치값을 담당하는 Position 객체에게 까지 전달하려면,
Cars -> Car -> Position 의 각 단계를 거쳐야 합니다.
Position에 대해서 올바르게 자동차가 이동했는지 테스트 코드를 작성하는 것은 당연하겠지만,
**Cars, Car에 대해서도 테스트를 작성해야할까? 에 대해서는 잘 모르겠습니다.**

가령, `cars.move(movementStrategy);` 와 같이 자동차들에게 전진을 지시하는 메소드 `move()`에 대해서
테스트 코드를 어떻게 작성해야 할지 감도 잘 오지 않습니다 ㅠ

### 3. TDD 코드의 작성 순서

TDD 방식으로 코드를 작성할 때, **Top-Down으로 작성해야할지, 혹은 Bottom-Up으로 작성해야할지?** 가 의문입니다.

지금까지 TDD가 아닌 일반적인 방식으로 코드를 작성한다면, Top-Down 방식으로 작성했었는데요. 
가장 상위에서부터 추상화 -> 구체화 과정을 거쳐나가면서 코드를 작성하는 것이 제게는 자연스러운 흐름이었는데,
TDD를 적용하려고 하는 순간 턱 막혀버렸습니다.

왜냐하면, Top-Down 방식으로 TDD를 진행할 경우 테스트 코드를 통과시키기 위해서는
너무 많은 Scope를 다뤄야하고, 그러다보니 변경점도 많아져서 오히려 관리하기가 어렵다는 생각이 들었습니다.

하지만 반대로 Bottom-up으로 진행할 때에도, TDD Cycle 을 따르는 것은 어렵지 않았으나,
"객체는 섬이 아니다" 라는 말과 같이, 서로간의 상호작용을 고려하는 순간 여러 문제점이 발생했습니다.
작성하다보니 기존 메소드 시그니쳐가 변경되어야 한다던지, 테스트 코드 자체가 변경되어야 한다던지..

교육때 예시로 제공되었던 StringCalculator의 경우 하나의 객체이고, 상호작용을 고려하지 않았기 때문에
어렵지 않게 연습할 수 있었는데, 규모가 조금만 생기더라도 위와 같은 문제점들이 발생하기 시작했습니다.
그러다보니 어디서부터 어떻게 시작해야할지 감이 잘 오지 않습니다.. ㅠㅠ 

쓰다보니 질문이 모호하고 장황해졌습니다. 글쓰기 실력이 서툰 점 양해 부탁드립니다. (__)
감사합니다!

cjlee38 committed:
test: String 클래스 split 학습 테스트 작성

1. 콤마로 분리해서 배열 반환
2. 콤마가 없는 경우 하나의 값을 가진 배열 반환

cjlee38 committed:
test: String 클래스 substring 학습 테스트 작성

cjlee38 committed:
test: String 클래스 charAt 학습 테스트 작성

1. 문자열로부터 인덱스 값을 기반으로 문자 하나를 가져온다.
2. 문자열의 인덱스 값이 범위를 벗어나는 경우 예외를 던진다.

cjlee38 committed:
test: Set 클래스 size 학습 테스트 작성

cjlee38 committed:
test: Set 클래스 contains 학습 테스트 작성

cjlee38 committed:
test: Set 클래스 contains & ParameterizedTest 학습 테스트 작성

cjlee38 committed:
feat(StringCalculator): 빈 문자열 또는 null 인 경우 0을 반환

cjlee38 committed:
feat(StringCalculator): 숫자 하나를 문자열로 입력할 경우 해당 숫자를 반환

cjlee38 committed:
feat(StringCalculator): 콤마로 구분된 여러 숫자의 합을 반환한다.

cjlee38 committed:
feat(StringCalculator): 콜론 혹은 콤마로 구분된 숫자의 합을 반환한다.

cjlee38 committed:
feat(StringCalculator): "//" 와 "\n" 문자 사이에 커스텀 구분자를 지정할 수 있다.

cjlee38 committed:
feat(StringCalculator): 숫자에 음수를 전달할 경우 예외를 던진다.

cjlee38 committed:
feat(StringCalculator): 공백이 주어질 경우 0을 반환한다.

Seungpang committed:
style: test코드 필요없는 public 제거

Seungpang committed:
docs(README): 설계 및 기능 목록 추가

Seungpang committed:
feat(InputView): 자동차 이름을 입력받는 기능 구현

Seungpang committed:
feat(Name): 자동차에 이름에 공백일 경우 예외 발생

Seungpang committed:
feat(Name):  자동차 이름은 5자 초과인 경우 예외 발생

Seungpang committed:
feat(Position): 자동차 위치 생성시 초기 값 0으로 지정

Seungpang committed:
feat(Car): 자동차를 생성하는 기능 구현

cjlee38 committed:
feat(Cars): 자동차의 이름을 ,(쉼표) 로 구분한다.

cjlee38 committed:
docs(README): 기능구현에 따른 커밋 및 설계 추가

cjlee38 committed:
feat(InputView): 시도할 회수(라운드)를 입력받는 기능 구현

cjlee38 committed:
feat(Round): 라운드에 시도할 회수 부여 구현


양수가 아닌 경우에 대한 예외처리 포함

cjlee38 committed:
feat(Movement) : 난수 기반 자동차 전진 전략 구현

cjlee38 committed:
feat(Car): 전략 기반 자동차 전진 구현

cjlee38 committed:
feat(OutputView): 자동차들의 위치를 라운드마다 표시

cjlee38 committed:
feat(OutputView): 최종 우승자를 출력


여러 명이 존재할 경우 ,(쉼표)로 구분하여 출력

cjlee38 committed:
refactor(OutputView): 자동차 잘못된 출력(줄바꿈) 제거

cjlee38 committed:
feat(Application): 레이싱 게임 시작 로직 구현

cjlee38 committed:
refactor(OutputView): 잘못된 출력 수정


1. 실행 결과 별 줄바꿈 추가
2. 실행 결과 한번만 출력되도록 변경
3. 최종 우승자 출력 메소드 시그니쳐 변경

cjlee38 committed:
refactor: 메소드 분리 및 가독성 리팩토링

Seungpang committed:
style: 구글 컨벤션에 맞도록 수정

cjlee38 committed:
docs(README): 프로그래밍 요구사항 점검

cjlee38 committed:
test(Car): 자동차 위치 확인 테스트 추가

cjlee38 committed:
test(Movement): 자동차 전진 전략 테스트 추가

lowoon commented on src/main/java/racingcar/domain/round/Round.java:
```diff
@@ -0,0 +1,42 @@
+        }
+    }
+
+    private static int parseInt(String round) {
```
static으로 만든 이유가 있을까요??

cjlee38 replied:
this() 를 사용하기 위해서 static으로 만들었습니다..만 util의 성격이 강한 것 같아 따로 분리하는 것이 좋을 것 같습니다.

lowoon commented on src/main/java/racingcar/controller/RacingCarController.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    private RacingCarService createRacingService() {
+        String carNames = inputView.inputCarNames();
```
view는 사용자의 UI를 담당하는 부분인데 이 부분이 controller 내부까지 들어와야 할까요??
입력이 inputView 형태가 아닌 Http요청과 같이 다른 형태로 들어올 때도 쓸 수 있나요??
서버의 역할을 좀 더 고민해보고, controller와 service를 만드셨는데 controller와 service의 역할에 대해서도 고민해보고 사용하시면 더 좋을거 같아요.

첫 미션인데 많은 부분이 반영되어 있어서 놀랐습니다. 이런 화두를 벌써 던져드리게 됐네요😳

cjlee38 replied:
service는 비즈니스 로직이 수행될 수 있는 행위(메소드)를 갖고 있고, 이를 적절하게 조합하는것이 controller의 역할이라고 생각합니다.  만약 웹서비스였다면 controller에 input view가 직접 들어오지 않고, view가 받아낸 DTO를 controller로 넘겨주는 것이 자연스러운 흐름일 것이라고도 생각을 하구요. 그런데 현재 자동차 레이싱 게임은 하나의 요청에 한번의 대응을 하는 웹 서비스와 달리 한번의 요청(자동차 이름 & 라운드) 에 대응하여 여러 번의 return을 해야 한다는 차이점이 있는 것으로 보입니다.(각 전진시도마다 실시간으로 자동차의 위치를 보여줌) 

따라서 while문 내에서 service 를 반복 호출하고, 그 결과를 반복해서 view로 넘겨주어야 하기 때문에 service와 view가 같은 위치에 있어야 하고, 그 위치를 controller로 결정했습니다. 피드백을 반영하여, 지난 프리코스 때 작성했던 것 처럼 Game이라는 클래스(혹은 단계)를 추가하여 Game에서 입력을 받아 controller로 넘겨주도록 변경하였는데요.  이렇게 작성할 경우 controller는 service를 조합하기보다, 단순히 service로 넘겨주는 역할만 하는 것처럼 보여 약간 껄끄럽게 느껴집니다. 

로운의 생각은 어떠신가요?

p.s. Http 요청이라는 상황을 가정하여 Client라는 scope를 새로 추가했습니다.

lowoon replied:
작업을 하시기 전에 이미 많은 고민 후에 나온 결과물이였군요! 👍 

**답변은 제 개인적인 생각인 것을 먼저 알려드려요!**

- 게임
기본적으로 서버는 app, web 등 프론트의 형태에 상관없이 동일한 결과를 보내야 하는 것을 항상 고려해야 하기때문에 view에 대해서 말씀을 드린 것이고요. 말씀하신 이유를 생각하면 '일반적인 서버 요청이 아니기 때문에 차리처럼 작업을 할 수 있겠다'라고 생각합니다!
다만, 제가 차리와 같이 생각을 했다면 라운드의 결과에 대한 값들을 합친 결과값을 만들어 view에 넘겨주고, view에서 모든 결과를 다보여줄지 1개만 보여줄지 정하게 했을거 같아요. 요청은 한번만 들어온다고 볼 수 있으니까요.
핸드폰 실시간 위치정보라고 했을 경우, 현재 위치만 보여줄 수도 있고, 지나온 길을 표시해줄 수도 있을텐데, histroy성으로 넘겨 front에서 정할 수 있도록 하는 방식이랄까요??

- 역할
service의 메서드를 적절히 조합한다는 것의 범위가 어떻게 될지 애매할 거 같은데요. 이 조합한다는 것이 비지니스 로직을 어느정도 포함하는 것일까요?? 그리고 controller의 역할을 껄끄럽게 느끼는 이유가 controller의 역할이 작기 때문인가요 아니면 역할에 맞지 않기 때문인가요?? 
제가 생각하는 controller의 역할은 들어온 request의 기본적인 검증(null check, empty check, authorization, header 등)을 하고 넘어온 데이터를 변환하여 service에 넘겨준 후, service로부터 받은 데이터를 응답의 형태에 맞게 변환하여 전달하는 것이라고 생각하고 있어요. (역할이 비대하여 분리된 경우, 기본적인 검증과 DTO 변환)
mvc를 보면 service가 없습니다. 이후에 service가 생기게 되었는데, 그 이유가 controller가 검증, 변환, 비지니스 로직 등 역할이 너무 많아 비대해졌기 때문에 역할과 기능을 분리하기 위해 나왔다고 생각해요. 지금과 같이 기능이 작은 경우에는 오히려 service를 안쓰는게 로직을 파악하기 쉽고 과도하지않은 분리일 수도 있지 않을까요??

위의 생각은 저만의 기준, 생각이었고요.
'대부분 이러한 형태로 하기 때문에 똑같이 한다'보다 왜 이러한 형태가 나오게 되었는지를 고민해보는게 추후의 작업에 대해서 자신만의 기준을 세울수 있다고 생각해요. 여러 방면으로 생각을 해보고 고민 후에도 지금과 같이 껄끄럽게 느껴지고, controller의 역할에 대해서 차리의 생각과 기준이 있다면 차리의 의견도 맞다고 생각합니다! 정답은 없으니까요!

제 답변이 차리가 고민했던 부분에 도움이 되었으면 좋겠네요.
혹시나 제가 잘못 이해한 부분이 있거나 이 부분이 아닌 다른 부분적으로 알고 싶으셨다면 DM이나 댓글을 통해 조금 더 얘기해봐요 😄

cjlee38 replied:
아직 배움이 짧아 말씀하신 내용을 다 흡수한 것 같지는 않습니다 😂

제가 이해한 내용을 바탕으로 정리를 해보자면

- 게임
: 현재 로직상으로는 사실 1번에 모아서 보여주나, 실시간으로 전달해서 보여주나 사람으로서는 차이를 거의 느낄 수 없는데요. 저는 게임이라는 단어에 얽매여서 그런지, 저는 "실시간으로 보여준다" 라는 점에 포인트가 꽂혀 있었던 것 같습니다. 그러다 보니, "만약 자동차가 1초마다 전진한다는 가정이 추가된다면 어떻게 될까?" 와 같은 상상을 하다 보니 위와 같이 작업했던 것 같습니다.

결국, 논점은 “작성한 app이 게임앱으로서 동작할거냐, 혹은 웹앱으로 동작할거냐” 에 대한 고려가 빠졌기 때문에 발생한 것 같다고 생각합니다. 그리고 당연하게도, 저희는 웹을 다루다보니 후자, 즉 로운이 말씀하신대로 한번에 넘겨주는 방법이 조금 더 타당하다는 생각이 듭니다!

- 역할
    
    : 말씀하신 내용이 제가 고민했던 부분의 허를 찌르시는 것 같습니다 ㅎㅎ;; 조금 더 구체적으로 말씀드리자면, 사실 controller - service - repository 로 이어지는 layered architecture와 model view controller의 MVC pattern 에 대해서 명확하게 이해하지 못하는 것 같기도 하구요.
    
    제 나름대로는 layered architecture와 MVC pattern의 조합은 일종의 식당으로 비유해서 생각하고 있었는데요. 컨트롤러는 홀에서 주문을 받는 매니저이고, 서비스는 요리사, 도메인은 요리 재료 ... 이런식으로요. 제가 말씀드렸던 “컨트롤러는 서비스를 적절히 조합한다" 와 같은 이야기는 마치 식당에서 세트 메뉴 주문을 받아 요리사에게 전달하는 것을 상상했던 것이구요. 마치 비유하자면 요리사서비스는 `cookRamen`, `cookKimbab` 메소드를 가지고 있고, 홀매니저컨트롤러는 `orderRamenSet` 와 같은 느낌으로요. 그래서 결국 `맥락을 주체하는 책임은 컨트롤러에게 있다` 와 같이 생각했는데, 이 부분이 많이 사라지다보니 controller의 역할이 퇴색되는 느낌이 들었던 것 같습니다.
    
    그리고 이러한 껄끄러운 느낌은 로운께서 짚어주셨던 대로 `과도한 분리` 로 인해 발생된 것 같습니다. 객체와 계층을 필요한대로 만들지 않고 일단 생각없이 Controller Service .... 이러한 네이밍을 붙인 부분이 틀렸던 것 같네요.
    

혹시라도 제가 잘못 이해했거나 짚고 넘어가시고 싶은 부분 있으시면 말씀해주시면 감사하겠습니다!

lowoon replied:
차리는 처음부터 많은 내용을 알고 계셨어요! 모든 작업을 많이 고민한 이후에 하신 것이라는 걸 얘기를 통해 느꼈어요. 💯 
고민을 통해 충분히 흡수하고 계십니다 😄 
`맥락`이 제가 생각하는 것과 차리가 생각하는 것이 다를 수 있을 것이라고 생각하는데요. **제 개인적인 생각**을 답변드릴게요.

저는 차리가 말한 `맥락을 주체하는 책임`이 service에 있다고 생각해요. controller까지는 presentation layer라고 생각되기 때문에 비지니스로직이 어떻게 동작하는 것인지 알 필요 없이 특정 요청에 대한 특정 응답만 주고, service가 비지니스 로직을 담당한다고 생각하는데요. service가 단순히 Repository에 crud만 하는 것이 아닌 어떠한 경우에 저장을 할지 업데이트 할지, 또는 업데이트를 하더라도 A 데이터는 반영하고 B 데이터는 반영을 안할지, crud 기능이 없이 메세지를 보낼지 말지 등 우리가 고객에게 제공하는 서비스에서 정한 정책에 따라 동작하도록 하는 것이라고 생각해요.
차리가 표현한 대로라면 요리사서비스는 `cook`이라는 메서드를 제공하고, 메뉴명과 type을 함께 넘겨주면 요리사가 어떻게 요리하는지는 모르지만 요리는 나오게 되는거죠. 
controller에서 set메뉴를 알아서 두개의 메뉴를 주문하는 것이 아니라 cook으로 메뉴를 넘기면서 set메뉴인지 일반메뉴인지 type을 함께 넘기게 되면 service가 type과 메뉴를 보고 필요한 요리를 만들어서 제공하는거죠.
일반 음식점에서는 메뉴가 잘못나오면 홀매니저가 주문을 잘못 넣은것으로 생각하지만 서버에서는 요리사가 잘못만들었다고 생각하게 되는 구조랄까요??

> A Service Layer defines an application’s boundary [Cockburn PloP] and its set of available operations from the perspective of interfacing client layers. It encapsulates the application’s business logic, controlling transactions and coor-dinating responses in the implementation of its operations.

https://tecoble.techcourse.co.kr/post/2021-04-25-dto-layer-scope/

차리가 말한 것을 제가 제대로 이해하고 답변했는지 모르겠네요 😅
제가 잘못 이해했거나 제가 말한 것중에 이해가 되지 않는 부분이 있으면 말씀해주세요

cjlee38 replied:
귀한 피드백 감사드립니다 👍.

아무래도 제가 제대로 공부하지 않고 controller의 역할에 대해서 이해한 것 같다는 생각이 강하게 드네요 ㅎㅎㅎ; 보내주신 링크도 읽어보니 한번 쯤은 제대로 짚고 넘어갈 포인트인 것 같네요. 덕분에 큰 배움을 얻는 것 같습니다.

감사합니다!!

lowoon reviewed (changes requested):
반가워요 차리! 리뷰어 로운입니다! 😄 
작성하신 코드 잘 확인했습니다. 
첫 미션임에도 너무 잘 구현하셨고 많은 부분이 구현되어 있어 놀랐네요. 👍 

질문하셨던 부분 답변을 드리면
1. domain
하나의 클래스가 하는 일의 많고 적음보다는 이 클래스의 역할을 무엇으로 했는지가 중요할 거 같아요. 역할에 맞는 동작을 하기 위해서 '상태와 행위를 한 곳에서 관리한다'가 필요하다고 생각하거든요.
예를 들어주신 cars의 역할이 car의 생성 및 관리라고 한다면 car를 생성하는 데 필요한 데이터와 검증은 cars의 역할에 부합하고 필요하기 때문에 문제가 되지 않을 거라고 생각합니다!
그리고 cars가 car들을 가지고 있기 때문에 가지고 있는 car들의 관리를 위해 업데이트 해주는 메서드를 제공하는 것 뿐이지, 로직의 흐름 중에 해당 메서드를 호출하는 service가 비지니스 로직을 담당하고 있는 것으로 봐야 하지 않을까요??

2. 테스트
name과 position의 테스트를 생성한다는 가정하에 말씀하신 car생성의 경우에는 값을 할당만 한다고 생각되어 개인적으로는 안해도 괜찮다고 생각합니다. 매개변수로 넘겨준 값이 잘 할당되었냐고 보는 것과 같으니까요. 단, name과 position을 직접 할당하는 것이 아닌 string을 매개변수로 넘기는 생성은 테스트를 해야하지 않을까요??
'어떠한 로직이 들어간다면 테스트를 하는게 좋다'고 생각하기 때문에 cars와 car의 경우는 테스트를 작성해야 한다고 생각합니다. 저를 포함 개발자들이 '이 부분은 확실하다'라고 생각했던 부분에서 오류가 발생하는 상황을 많이 겪거든요.
테스트 방법은 random값으로는 테스트가 어려우니 상수를 입력했을 때 제대로 동작하는지 테스트하면 되지 않을까요??

3. TDD 
전체적인 흐름을 아는 채로 TDD를 하는 것은 어렵다고 생각합니다. 그래서 항상 가장 작은 단위, 가장 작은 기능부터 시작하는 것 같아요. 작은 단위의 테스트를 작성하는 것은 그렇게 어렵지 않으니까요. 하지만 말씀하신 것처럼 작은 단위의 기능을 만든 후에 수정이 필요한 상황이 생길수 밖에 없어요. 그럼 그때 메서드와 테스트를 수정하면 되지 않을까요?? 처음부터 완벽한 기능, 메서드를 만들 수 없고, 말씀하신 것처럼 다른 객체와의 관계, 로직의 변경 등을 고려하면 수정사항이 생기는 것은 당연한 일이니까요. 그렇기 때문에 수정이 필요할 때 기능과 테스트를 함께 수정하는 것도 TDD의 과정이라고 생각합니다. 

답변이 도움이 되었으면 좋겠네요 😸 
궁금한 점이 생기면 언제든지 DM 혹은 댓글남주세요 ^-^

lowoon commented on src/main/java/racingcar/domain/car/Cars.java:
```diff
@@ -0,0 +1,68 @@
+
+public class Cars {
+
+    private static final int MIN_LENGTH = 0;
```
차량들의 최소 개수의 의미인거 같은데 좀 더 명확한 변수명으로 수정해주는게 낫지 않을까요?

cjlee38 replied:
동의합니다. MIN_LENGTH라는 이름은 너무 모호한 것 같습니다.

lowoon commented on src/main/java/racingcar/domain/car/Cars.java:
```diff
@@ -0,0 +1,68 @@
+        return max;
+    }
+
+    public void move(MovementStrategy strategy) {
```
주요 로직 메서드는 getter보다 위에 있는게 좋은거 같아요

```
public class ClassName {
	// private static final 
	// 상수
	// 클래스 변수
	// 인스턴스 변수
	// 생성자
	// 팩토리 메서드
	// 일반 메서드
	// getter, setter
	// equals, hashCode, toString
}
```

cjlee38 replied:
반영했습니다!

lowoon replied:
반영이 안된거 같아요 👀

cjlee38 replied:
우승자를 찾는 메소드를 getter 메소드라고 착각했습니다 😢 다시 수정했습니다.

cjlee38 committed:
refactor(Round): parseInt 메소드를 utils로 추출

cjlee38 committed:
refactor(Cars): 반복문 stream으로 수정

cjlee38 committed:
refactor(Cars): 변수명이 명확한 의미를 갖도록 수정


자동차의 최소 개수(MIN_LENGTH -> MINIMUM_CAR_QUANTITY)

잘못된 입력으로 인해 자동차의 개수가 0개인 경우 예외
CarNamesNotOnlyCommaException -> InvalidNameInputException

cjlee38 committed:
refactor(Cars): 로직 메소드 위치 변경

cjlee38 committed:
refactor(Car): 객체 생성 시 null 검증 로직 추가

cjlee38 committed:
refactor(Round): 종료 조건 여부 책임을 round에 할당

cjlee38 committed:
refactor: 메소드 네이밍 변경

cjlee38 committed:
feat(Client): 사용자 입력을 관리하는 client scope 추가

cjlee38 committed:
refactor(controller): controller와 view 분리


Controller와 View를 handle 하는 User의 역할인 Game 클래스를 추가

cjlee38 committed:
feat(Position): 자동차 위치 검증로직 추가

lowoon commented on src/test/java/racingcar/domain/car/CarsTest.java:
```diff
@@ -0,0 +1,30 @@
+        String input = "pobi,seung,char";
+        Cars cars = new Cars(input);
+        List<Car> carList = cars.getCarList();
+        Assertions.assertThat(carList.size()).isEqualTo(3);
```
assertAll을 적용해보는 것도 좋을 거 같네요

cjlee38 replied:
확인해보니 assertAll 은 assertj Assertions 에는 안보이고 Junit Assertions 에 있는 것 같습니다!

여담이지만 junit 과 assertj 두 가지를 선택하는 기준은 개인 기호나 팀의 스타일에 따르는 편인가요?
혹은 저마다의 장단점이라고 볼만한게 있어서 상황마다 선택하는 기준같은 것이 있을까요?

lowoon replied:
팀마다 다를거 같기는 한데 기본적으로 assertj를 사용하는데 junit에서 제공하는 것을 사용 안하지는 않아요.
하나의 테스트 안에서 한가지만 정해서 사용할 필요없이 필요하면 함께 쓰는것 같구요.
```
assertAll(
        () -> Assertions.assertThat(carList.size()).isEqualTo(3),
        () -> Assertions.assertThat(carList.get(0)).isEqualTo(new Car("pobi"))
);
```
assertAll을 말씀드린건 assertThat을 여러번 쓸때 첫 assertThat 구문에서 테스트가 실패할 시 밑에 있는 assertThat은 돌지 않는데 assertAll 안에서는 첫 assertThat이 실패하더라고 밑에 있는 모든 assertThat을 돌려봐서 입니다.

cjlee38 replied:
아하.. 이렇게 혼용할 수도 있군요! 하나만 써야한다는 고정관념이 있었네요.. 좋은 팁 감사드립니다!

lowoon commented on src/main/java/racingcar/view/OutputView.java:
```diff
@@ -0,0 +1,49 @@
+        outputClient.output("\n");
+    }
+
+    public void printWinners(List<Car> carList) {
```
이미 알고 계신거 같은데 도메인을 직접 view에 안넘기도록 해볼까요??

cjlee38 replied:
말씀하신 내용은 아마 domain(model) 대신 DTO를 view에 넘겨주는 것..이 맞겠죠 ㅎㅎ..?

그런데 작성하다보니 드는 의문점이, view에 CarsDTO를 넘겨줄 수도 있고, List<CarDTO> 를 넘겨줄 수도 있는데요.
리스트로 넘겨주는 편이 조금 더 유연하게(단일 CarDTO를 넘겨주는 경우를 상정) 넘겨줄 수 있을것 같아서 이를 선택했는데,
CarsDTO 를 사용했을 때 장점이 있을까요 ?

그리고 두 번째는, Model 대신 DTO를 넘겨주는 이유는 Model을 보호하기 위함이라고 알고는 있는데(아마 여기서 보호한다는 것은 객체의 내부 값이 변경될 가능성을 이야기 하는 것이겠죠..?), Car 내부에는 Name 클래스와 Position 클래스가 들어있는 상황인지라, Car를 CarDTO로 변환할 때 Name과 Position을 필드로 각각 갖는 것은 좀 이상하다고 생각했습니다. Name과 Position 자체도 inplace로 수정될 수 있는 가능성이 있으니까요. 그러다보니 CarDTO 를 생성할 때 immutable한 String과 int를 넘겨주는 것이 맞다고 판단했습니다. 다른 방법으로는 NameDTO, PositionDTO 를 추가하는 것인데 이는 time cost가 상대적으로 높다고 생각해서 사용하지 않았구요.. 

로운이라면 어떤 선택을 하셨을 지 궁금합니다!

lowoon replied:
네 말씀하신 DTO를 넘겨주는 것이 맞고요 😄 
1. 현재는 필요한 데이터만 넘겨도 되기 때문에 List<CarDTO>여도 상관이 없을텐데요. 실제로는 필요한 데이터만 넘기지 않고, timestamp, code, message, data 등 여러 값을 front와 논의 후 정해서 보내고 있어요. 이렇게 필요한 데이터(게임의 결과값)뿐만이 아니라 다른 값들이 있을때는 CarsDTO가 편할 수 있겠죠??
2. 두번째는 저도 차리처럼 했을거 같아요. 실제로 필요한 값이 String과 int인데 이걸 굳이 한번더 감쌀 필요는 없다고 생각합니다. 이미 이전에 객체를 생성하면서 검증은 통과했고 dto에서 필요한 것은 값일 뿐이니까요.

lowoon reviewed (approved):
고생하셨어요. 차리! 이전 리뷰 내용을 잘 반영해주셨습니다 💯
1단계에서 이미 많은 부분이 반영되어 있었고, 많은 부분을 고민했다는 걸 알았어요! 👍 
몇가지 코멘트를 달아봤는데요. 코멘트 반영은 2단계에서 부탁드릴게요.
추가적인 궁금증은 이 pr이 머지되어도 댓글 남겨주셔도 좋구 DM도 좋습니다.

lowoon commented on src/main/java/racingcar/domain/car/Cars.java:
```diff
@@ -0,0 +1,65 @@
+            .collect(Collectors.toList());
+    }
+
+    public void move(MovementStrategy strategy) {
```
가장 중요 로직인데 테스트가 없는것 같네요!

cjlee38 replied:
추가했습니다!

lowoon commented on src/main/java/racingcar/domain/round/Round.java:
```diff
@@ -0,0 +1,38 @@
+
+    private static final int END_ROUND = 0;
+    private static final int MIN_NUMBER = 1;
+    public static final String ERROR_MESSAGE = "1 이상의 양수를 입력하세요.";
```
접근제어자가 public이네요?? 이유가 있을까요?

cjlee38 replied:
실수로 놓쳤습니다 ㅠ private 으로 수정했습니다.
